File: ProfileManager.cpp
#include "ProfileManager.hpp"
#include <cmath>
#include <stdexcept>

// Инициализация статического поля
int ProfileManager::totalProfiles = 0;

// Конструктор: Инициализирует все поля класса
ProfileManager::ProfileManager(const std::string& id, const std::string& username, int age, Gender gender, int heightCm,
    double currentWeightKg, Goal targetGoal, Level fitnessLevel,
    const std::list<Weekday>& workoutDaysPref, const std::set<Equipment>& availableEquipment,
    bool alertsEnabled)
    : id(id), username(username), age(age), gender(gender), heightCm(heightCm), currentWeightKg(currentWeightKg),
    targetGoal(targetGoal), fitnessLevel(fitnessLevel), workoutDaysPref(workoutDaysPref),
    availableEquipment(availableEquipment), alertsEnabled(alertsEnabled) {
    totalProfiles++;
}

// Конструктор копирования
ProfileManager::ProfileManager(const ProfileManager& other)
    : id(other.id), username(other.username), age(other.age), gender(other.gender),
    heightCm(other.heightCm), currentWeightKg(other.currentWeightKg), targetGoal(other.targetGoal),
    fitnessLevel(other.fitnessLevel), workoutDaysPref(other.workoutDaysPref),
    availableEquipment(other.availableEquipment), alertsEnabled(other.alertsEnabled) {
    totalProfiles++;
}

// Деструктор
ProfileManager::~ProfileManager() {
    totalProfiles--;
}

// Метод: Обновляет текущий вес
void ProfileManager::changeWeight(double newWeight) {
    if (newWeight <= 0) {
        throw std::invalid_argument("Вес должен быть положительным");
    }
    currentWeightKg = newWeight;
}

// Метод: Обновляет цель
void ProfileManager::updateGoal(Goal newGoal) {
    targetGoal = newGoal;
}

// Метод: Рассчитывает BMI по формуле вес / (рост^2)
double ProfileManager::calculateBMI() const {
    try {
        if (heightCm <= 0) {
            throw std::runtime_error("Рост должен быть положительным");
        }
        double heightM = heightCm / 100.0;
        return currentWeightKg / (heightM * heightM);
    }
    catch (const std::exception& e) {
        throw;
    }
}

// Перегрузка оператора ==
bool ProfileManager::operator==(const ProfileManager& other) const {
    return this->id == other.id && this->username == other.username;
}

// Перегрузка оператора <
bool ProfileManager::operator<(const ProfileManager& other) const {
    return this->username < other.username;
}

// Дружественная функция
std::ostream& operator<<(std::ostream& os, const ProfileManager& profile) {
    os << "Профиль: " << profile.username << " (ID: " << profile.id << ")";
    return os;
}

File: FitnessAppCsharp/ProfileManager.cs
using System;
using System.Collections.Generic;

public class ProfileManager
{
    // Перечисления для характеристик пользователя
    public enum Gender { Male, Female }                         // Пол пользователя
    public enum Goal { LoseWeight, GainMass, Health }           // Цель тренировок
    public enum Level { Beginner, Intermediate, Advanced }      // Уровень подготовки
    public enum Weekday { Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday } // Дни недели
    public enum Equipment { Dumbbells, Barbell, Bench }         // Доступное оборудование

    // Статическое поле для подсчета количества созданных профилей
    private static int profileCount = 0;
    // Статический метод для получения общего количества профилей
    public static int GetProfileCount() => profileCount;

    //Поля класса 
    private string id;                                          // Уникальный идентификатор пользователя
    private string username;                                    // Имя пользователя
    private int age;                                            // Возраст пользователя в годах
    private Gender gender;                                      // Пол пользователя
    private int heightCm;                                       // Рост в сантиметрах
    private double currentWeightKg;                             // Текущий вес в килограммах
    private Goal targetGoal;                                    // Цель тренировок
    private Level fitnessLevel;                                 // Уровень подготовки
    private List<Weekday> workoutDaysPref;                      // Список предпочитаемых дней для тренировок
    private List<Equipment> availableEquipment;                 // Список доступного оборудования
    private bool alertsEnabled;                                 // Включены ли уведомления

    // Конструктор: инициализирует профиль пользователя
    public ProfileManager(string id, string username, int age, Gender gender,
        int heightCm, double currentWeightKg, Goal targetGoal, Level fitnessLevel,
        List<Weekday> workoutDaysPref, List<Equipment> availableEquipment,
        bool alertsEnabled)
    {
        // Использование this 
        this.id = id;
        this.username = username;
        this.age = age;
        this.gender = gender;
        this.heightCm = heightCm;
        this.currentWeightKg = currentWeightKg;
        this.targetGoal = targetGoal;
        this.fitnessLevel = fitnessLevel;
        this.workoutDaysPref = workoutDaysPref ?? new List<Weekday>();
        this.availableEquipment = availableEquipment ?? new List<Equipment>();
        this.alertsEnabled = alertsEnabled;
        profileCount++;
    }

    public string Username
    {
        get => username;
        set => username = value;
    }

    public double GetCurrentWeightKg() => currentWeightKg;
    public int GetHeightCm() => heightCm;
    public Level GetFitnessLevel() => fitnessLevel;

    // Метод изменения веса с обработкой исключений
    public void ChangeWeight(double newWeight)
    {
        try
        {
            if (newWeight <= 0) throw new ArgumentException("Вес должен быть положительным числом");
            this.currentWeightKg = newWeight;
        }
        catch (ArgumentException ex) 
        {
            Console.WriteLine($"Ошибка при обновлении веса: {ex.Message}");
        }
    }

    // Метод расчета индекса массы тела с обработкой исключений
    public double CalculateBMI()
    {
        try
        {
            if (this.heightCm <= 0) throw new InvalidOperationException("Рост должен быть положительным для расчёта BMI");
            double heightM = heightCm / 100.0;
            return Math.Round(currentWeightKg / (heightM * heightM), 2);
        }
        catch (InvalidOperationException ex) 
        {
            Console.WriteLine($"Ошибка расчёта BMI: {ex.Message}");
            return 0.00;
        }
    }

   
    public List<Equipment> GetAvailableEquipment() => availableEquipment;
}
